library/ssl_msg.c

```diff
@@ -4554,6 +4565,41 @@ static int ssl_buffer_future_record( mbedtls_ssl_context *ssl,
 
 #endif /* MBEDTLS_SSL_PROTO_DTLS */
 
+/*
+ * RFC 8446:
+ * "If the client attempts a 0-RTT handshake but the server
+ *  rejects it, the server will generally not have the 0-RTT record
+ *  protection keys and must instead use trial decryption (either with
+ *  the 1-RTT handshake keys or by looking for a cleartext ClientHello in
+ *  the case of a HelloRetryRequest) to find the first non-0-RTT message."
+ */
+#if defined(MBEDTLS_SSL_SRV_C)
+static int ssl_should_drop_record( mbedtls_ssl_context *ssl )
+{
+#if defined(MBEDTLS_ZERO_RTT) && !defined(MBEDTLS_SSL_USE_MPS)
+    if( ssl->conf->endpoint != MBEDTLS_SSL_IS_SERVER || ssl->handshake == NULL )
+        return( 0 );
+
+    /*
+     * Drop record iff:
```

Comment:
I don't understand how the 4 cases are distinguished in the code below.

Reply:
`skip_failed_decryption` is set to `1` in one place if the client sends early data indication but server can't/won't support early data (1 and 2), and is set to `0` once a record is deprotected successfully (3). 

Checking `transform_in` ensures handshake traffic key was used when attempting to deprotect (4). I added this condition as an invariant check since there should never be a case where `skip_failed_decryption == 1` but we're using other keys.

I agree it's not immediately obvious, do you think the implementation could be more clear or just the comments?